### PR TITLE
Init backend

### LINKED ISSUE #6: Set up Backend Server

### PR DESCRIPTION
**Pull request by joshhn**
This PR initializes the backend using Node.js, Express with TypeScript. 

@kylepham's checklist:
- [x] Set up Express.js app with TS.
- [x] Expose a "hello world" endpoint to test application.
- [x] Create 2 MongoDB DB instances:
  - [x] 1 named `dev`: All development progress should be committed to this DB.
  - [x] 1 named `prod`: DB for production.
- [x] Connect to DB on server initiation.
- [x] Create a "toy" schema and test with one or more CRUD operations.
- [x] Set up `prettier` with the following rules in `.prettierrc.json`:
  ```ts
  {
    "tabWidth": 2,
    "semi": true,
    "singleQuote": false,
    "printWidth": 120
  }
  ```
- [x] Add `.gitignore`.
- [x] Add all DB credentials to `.env` and share internally.
- [x] Use [`nvm`](https://github.com/nvm-sh/nvm) to enforce consistent Node.js runtime, ideally Node.js v20.

### CODE REVIEW

File: src/service/database.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    const insertManyResult = await collection.insertMany(data);
+    console.log(`${insertManyResult.insertedCount} documents successfully inserted.\n`);
+    return true;
+  } catch (err) {
+    console.error(`Something went wrong trying to insert the new documents: ${err}\n`);
+    return false;
+  } finally {
+    await client.close();
+  }
+}
```
Comment (kylepham) [inline src/service/database.ts]: 1. Split functions into their designated files and folders.
2. Avoid connect to DB on function trigger. Should establish DB connection on server initiation.

File: src/service/database.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
+import { MongoClient, ServerApiVersion } from 'mongodb';
+import User from '../schemas/User'
+import Card from '../schemas/Card'
+
+const uri = "mongodb+srv://admin:adminpass@cluster0.povokmt.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0";
+const dbName = "myDatabase";
```
Comment (kylepham) [inline src/service/database.ts]: 1. Add credentials to .env file
2. Constant variable should be in uppercase.

File: src/schemas/Card.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+interface Card {
+  name: string;
+  price: number;
+  quantity: number;
+}
+
+export default Card;
```
Comment (kylepham) [inline src/schemas/Card.ts]: Add new line at EOF (format using `prettier`).

File: src/routes/users.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    } else {
+      res.status(500).json(data)
+    }
+    
+  } catch (error: any) {
+    res.status(400).json( { error: error?.message } )
+  }
+})
+
+export default router
```
Comment (kylepham) [inline src/routes/users.ts]: Add new line at EOF (format using `prettier`).

File: src/routes/cards.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    } else {
+      res.status(500).json(data)
+    }
+    
+  } catch (error: any) {
+    res.status(400).json( { error: error?.message } )
+  }
+})
+
+export default router
```
Comment (kylepham) [inline src/routes/cards.ts]: Add new line at EOF (format using `prettier`).

File: src/index.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  response.status(200).json({ message: "Bye World" });
+});
+
+// specify the api path for the server to use
+app.use('/users', usersRouter)
+app.use('/cards', cardsRouter)
+
+app.listen(PORT, () => {
+    console.log(`server listening on http://localhost:${PORT}`)
+})
```
Comment (kylepham) [inline src/index.ts]: Add new line at EOF (format using `prettier`).

File: tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    /* Experimental Options */
+    // "experimentalDecorators": true,        /* Enables experimental support for ES7 decorators. */
+    // "emitDecoratorMetadata": true,         /* Enables experimental support for emitting type metadata for decorators. */
+
+    /* Advanced Options */
+    "resolveJsonModule": true /* Include modules imported with '.json' extension */
+  },
+  "include": ["src/**/*"],
+  "exclude": ["node_modules/**/*", "build/**/*"]
+}
```
Comment (kylepham) [inline tsconfig.json]: Add new line at EOF (format using `prettier`).

File: .gitignore [hunk 8]
```diff
@@ -0,0 +1,7 @@
+node_modules
+
+.env
+
+**/DS_Store
+
+build
```
Comment (kylepham) [inline .gitignore]: Add new line at EOF (format using `prettier`).

File: src/config/dotenv.ts [hunk 9]
```diff
@@ -0,0 +1,3 @@
+import dotenv from 'dotenv'
```
Comment (kylepham) [inline src/config/dotenv.ts]: Redundant file?

**Review by kylepham — CHANGES_REQUESTED** [review]

**Comment by kylepham** [comment]
Updated the PR description with a checklist.

File: build/index.js.map
Comment (kylepham) [inline build/index.js.map:1]: Ignore `build` folder.

**Review by kylepham — CHANGES_REQUESTED** [review]

File: src/database/database.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+const DB_NAME = "admindb";
+
+const MONGODB_URI =
+  "mongodb+srv://" +
+  DB_USER +
+  ":" +
+  DB_PASSWORD +
+  "@" +
+  cluster_url +
+  ".povokmt.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0";
```
Comment (kylepham) [inline src/database/database.ts]: Use string literals syntax.

File: src/database/database.ts [hunk 11]
```diff
@@ -0,0 +1,52 @@
+import { MongoClient, ServerApiVersion } from "mongodb";
+
+const DB_USER = process.env.DB_USER as string;
+const DB_PASSWORD = process.env.DB_PASSWORD as string;
+const cluster_url = "cluster0";
```
Comment (kylepham) [inline src/database/database.ts]: Uppercase constant variable.

File: src/database/card.ts [hunk 12]
```diff
@@ -0,0 +1,20 @@
+import Card from "../schemas/Card";
+import { getMongoDBClient } from "./database";
+
+const dbClient = getMongoDBClient();
+const dbName = "carddb";
+const collectionName = "Card";
+
+export async function insertCard(data: Card[]) {
+  await dbClient.connect();
```
Comment (kylepham) [inline src/database/card.ts]: Connection to DB should always be open. Establish connection along with spinning up the server, and reuse DB instance on every API endpoint without reconnecting again. 
[Reference](https://www.mongodb.com/resources/languages/express-mongodb-rest-api-tutorial).

**Review by kylepham — CHANGES_REQUESTED** [review]

File: src/database/card.ts (hunk 12 above)
Comment (joshhn) [reply src/database/card.ts]: I already reuse the connected client for the whole server, but let me check out the suggested reference.

Comment (kylepham) [reply src/database/card.ts]: The same client instance inefficiently gets triggered to establish a connection again.

Comment (joshhn) [reply src/database/card.ts]: I see. I forgot to delete it.

**Review by kylepham — APPROVED** [review]
LGTM! 🚀

**Pull request merged**